sklearn/tree/_tree.pyx

```diff
@@ -446,7 +454,8 @@ cdef class BestFirstTreeBuilder(TreeBuilder):
 
         if not is_leaf:
             splitter.node_split(impurity, &split, &n_constant_features)
-            is_leaf = is_leaf or (split.pos >= end)
+            is_leaf = (is_leaf or split.pos >= end or
+                       split.improvement + EPSILON < min_impurity_decrease)
```

Comment:
To clarify further. Setting it to `1e-7` as done for other stopping params to denote `eps` will not let the tree grow fully and produce trees dissimilar to master...